### PR TITLE
Fixed build with CMake with MinGW-w64 and packaging for Windows

### PR DESCRIPTION
**Pull request by KOLANICH**
* Removed M$ EEExtensions from SAPI5 module in order to compile it with free toolchains.
* Implemented some CMake code to use some headers from Windows SDK missing in MinGW.
* Moved some implementations from header files into source ones.
* Splitted some code used by forked versions of hts_engine and libmage into separate libs to allow build everything as shared libs on Windows. *nixes don't have this issue.
* Improved hardening code.
* Improved packaging - added some text description and metadata for fancy installer build with NSIS CPack.
* Resolved some warnings detected by CLang 11
* .gitignored some files

Now about the issues and the instructions how to build using MinGW-w64.

1. Currently automatic registration and deregistration of SAPI 5 engine COM server is not implemented. One has to call COM server manually. In future it should be called by the installer.

2. Currently automatic registration and deregistration of SAPI 5 engine is not fully implemented. In fact it is implemented, but the CMake lib doing this was tested only for generating a reg file. I haven't tested if that works as intended for WIX (and it is not yet fully implemented for it, I am not yet enough familiar to WIX), I am unable to test that for NSIS for this project because of a bug in CPack NSIS backend, and I am not sure how to implement that properly for Qt IFW. So, the lib needs to be tested properly. For now what is tested is only generation of a reg file, and it still should be improved a lot. 

3. Currently there are some problems with clang 11. Sometimes it uses the wrong exception model, even if it is told to use the correct one. I.e. the latest official build of MinGW-w64 toolchain for 32-bit x86 is shipped in 2 variants, sjlj and dwarf exception models. I use sjlj because of compatibility to some prebuilt libs. I had to disable building of `RHVoice-make-hts-labels` for the problematic version of CLang with MinGW-w64 runtime.

4. MinGW-w64 from the latest build on SourceForge is not very compatible with PIC. You may want to disable PIC when building againt MinGW runtime.

5. SAPI 5 in MinGW is far from being complete. The solution is to use vanilla Windows SDK for needed headers. It is not that simple - the preshipped headers were made to be used with Visual Studio. The solution is to recompile the headers frok their IDLs with WIDL - a replacement for MIDL generating the code compatible to MinGW stdlib. I have created a CMake lib for this. It was not that simple as you can imagine. The most of the time spent on implementing this PR was spent on implementing and debugging that lib.
    * widl has lot of bugs. Some of them were fixed just recently. Some are not yet fixed. I had to work around them by patching the IDLs.
    * widl crashes on some IDLs, though produces the output we needec. This causes build to fail. I had to write a tool for calling other tools, that suppresses errors. It is not yet compatible to x64 (it'd be trivial to fix) and has some other issues. Also using it has a drawback - we should either have it prebuilt or we build it during build, but unable to cross-compile. The proper fix is to resolve all the memory-safety issues in widl.
    * You heed a bleeding edge version of widl. I have extracted it from wine source and some its dependencies and created CMake scripts to build them easily and cross-platformly. But its essential dependency is not properly extracted, I have to finish the tool for doing that first. Just had no time for that yet. So, now, to build the needed version of widl, you need git for windows with git bash, mingw-w64 and win_flex_bison (you will have to rename the binaries). Download a snapshot of wine from its github repo, unpack, go to that dir, get help on configure script, copy the help to a text editor, make a command line for the configure script, disabling all the features, configure and make (you will have to patch the makefile after configuration in order to set the right make binary). Stop after it starts compiling idl files and save the compiled widl somewhere.

6. The location of data and configs is hardcoded. We need to use location repative to the executable for Windows platform;  I haven't yet implemented that.

7. Hardening for Windows seems to be non-working (all the flags are detected as unsupported, though I know exactly they are supported

8. THE SAPI 5 MODULE CRASHES THE CONTROL PANEL APPLET WHEN ANY OF ITS VOICES IS SELECTED. IDK why this happens, it needs some investigation.

I want you all to review this PR and the users of VS - to check if any changes in this PR conflict to the building with VS. It worth to review not only the code in this repo, but also all the tools and libs I have written in process. They are added as submodules and their deps.

### CODE REVIEW

**Comment by alex19EP** [comment]
hello. i plan to look at it ASAP.

Do I understand correctly that the building with microsoft toolchain works?

**Comment by KOLANICH** [comment]
>Do I understand correctly that the building with microsoft toolchain works?

No. I don't have the Studio, so I haven't tested on it.

BTW, I have splitted this PR into multiple pretty standalone PRs. Some of them are ready to be merged.

**Comment by alex19EP** [comment]
thanks for splitting.

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,17 +14,25 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 
-cmake_minimum_required(VERSION 3.5.1)
+cmake_minimum_required(VERSION 3.9)
 project("RHVoice")
 
-set(CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}" "${CMAKE_SOURCE_DIR}/cmake")
+set(customCMakeModulesDir "${CMAKE_SOURCE_DIR}/cmake")
+list(APPEND CMAKE_MODULE_PATH "${customCMakeModulesDir}" "${customCMakeModulesDir}/thirdparty/MinGWSAPI5SDK" "${customCMakeModulesDir}/thirdparty/CPackRegistry")
```
Comment (msva) [inline CMakeLists.txt:21]: how about hiding this under if(mingw) or if(windows) too?
I bet it is not needed on the non-windows OSes

Comment (KOLANICH) [reply CMakeLists.txt:21]: Thanks, probably should do that.

**Comment by alex19EP** [comment]
hay @KOLANICH when you restart working on this, I suggest to rebase on master. so it can be checked by our CI.

**Comment by KOLANICH** [comment]
It is not ready yet. First I should fix the issues introduced (likely by me) somewhen this year in CMake scripts. The issue is that some voices are not packaged. Likely the issue was introduced in my patches from Windows series, I had fixed an error preventing packaging data for windows, but it may have disrupted (has this disrupted, or something else, it is not known yet) packaging of data for linux.

Then I need to land the patches reorganizing packaging. This PR probably should be landed the last of them all. I have also something to fix here, the most importantly the code untying resources from absolute paths (1. detecting the path of a dll, 2. looking up resources relative to the path of dll, not to cvd) because currently the sapi5 module tries to discover the resources using hardcoded paths. It is not that easy because the official builds of MinGW-W64 are pieces of shit in the sense of support of UCS-16 paths (which are musthave for windows, otherwise the install path won't be able to contain cyrillics) as used in Windows (I have tried to build an own mingw-w64 stdlib from sources without any success and I cannot recommend using non-official third-party builds because I don't trust them, especially after somone on StackOverflow has claimed that he has seen a build with malware), so some nasty workarounds have to be done.

**Comment by alex19EP** [comment]
ok. thanks for clarifying this.
regarding mingw. i am using msys2 which comes with it.

**Comment by abitrolly** [comment]
> the official builds of MinGW-W64 are pieces of shit in the sense of support of wchar_t paths

@KOLANICH if you are able to create an official issue about that, that will surely serve better to find a resolution than a rant on unrelated bug tracker. :D

**Comment by KOLANICH** [comment]
@abitrolly, these issues are [long known](https://sourceforge.net/p/mingw-w64/mailman/search/?q=wchar_t+stream)

**Comment by abitrolly** [comment]
That's a mailing list. I expect a tracker item for.. you know, tracking.

**Comment by KOLANICH** [comment]
1. the problem may be solved in upstream gcc. But MinGW-w64 has no recent official prebuilt releases.
2. I am not going to register on SourceForge.

**Comment by abitrolly** [comment]
> I am not going to register on SourceForge.

Well, I am sure they have [reasons](https://itsfoss.com/youtube-dl-github-takedown/) to take measures against. MinGW is on OSDN, with equally hardcore user experience. For your specific issue https://github.com/msys2/mingw-packages/issues may be the right place, though.

**Comment by abitrolly** [comment]
> MinGW-w64 has no recent official prebuilt releases.

GCC-10.2.0-crt-8.0.0-with-ada-20201019.7z from https://sourceforge.net/projects/mingw-w64/files/latest seems pretty recent to me.

**Comment by KOLANICH** [comment]
>`/Toolchains targetting Win64/Personal Builds/ray_linn/GCC-10.X-with-ada/GCC-10.2.0-crt-8.0.0-with-ada-20201019.7z`

It is not an official release build, it is a personal build by someone called Ray Linn

**Comment by KOLANICH** [comment]
>issue https://github.com/msys2/mingw-packages/issues may be the right place, though.

Maybe. IDK if I should use them or not.

**Comment by KOLANICH** [comment]
Hi. Why have all my PRs been closed without any description why they have been closed? Was it done automatically? Should they be reopened?

**Comment by KOLANICH** [comment]
I think that

1. The PRs automatically closed by the migration should be determined. It should be available through `events` REST API. One shoold just download enough of events and filter by time.
2. They should be automatically reopened with a script.

Could that be done, please?

**Comment by KOLANICH** [comment]
P.S. I'm going to rebase the PRs.

**Comment by alex19EP** [comment]
> I think that
>     2. They should be automatically reopened with a script.

unfortunately it is not possible to reopen these prs. I confess honestly, I did not take into account such a problem when rewriting history.

maybe when branch is rebased it can be reopened but for now button reads as: The KOLANICH-tools:cmake branch has no history in common with RHVoice:master.

**Comment by KOLANICH** [comment]
>I confess honestly, I did not take into account such a problem when rewriting history.

So did I. But even if I had known we will have such issues, I still think the rewriting worth. Cloning has become much faster.

**Pull request closed without merging**